rust/wal3/README.md

```diff
@@ -539,44 +580,40 @@ object store.  This becomes expensive as the number of logs grows, especially if
-potential log name and could enumerate them?  We could do a GET/HEAD on each dirty "bit" to see if
-it exists in object storage.  This comes out to $5k per pass for 1e9 objects, but violates the
-requirement that work be proportional to the data written.
+This means that we can track the dirty logs with a single 48B record per fragment.  Where to put
+that record?  We need some way to record them as they happen and then scan/roll-up the records into
+a summary of dirty logs at all times.
 
-The observation to have is that if collection dirty bits are 1:1 with files in object storage, then
-there will be O(dirty collections) files to list.  One straw man approach considered was to make a
-hierarchy in S3 like a B+-tree:
+Viewed another way, we have a stream of events that we need to record approximately in order.
 
-```text
-wal3/dirty/ea/8c/d5/d4/f1b2ed79d8415c832e2fd2cef87e6ebcecde12aa01b2ccc5aceaab55
-wal3/dirty/ea/47/84/3f/7d677b67d7364c2b34322770988e91e2b21df4d7c98064dfb14ae68d
-```
+What if we just put it in another log?  wal3 is already configured to dynamically batch and write
+log data to object storage in an efficient manner.  Further, we know a single machine's multiplicity
+of logs can be sustained by the throughput of a single log by the math above.
 
-This strictly increases the number of list operations, but brings a useful property:  We can
-partition the space of dirty bits so that multiple scanners can pass over them in parallel.  This
-doesn't reduce the cost of finding dirty logs, but does make it faster.
+The protocol for discovering dirty logs, then, is to write to a "dirty" log and roll-up the dirty
+log for compaction.  This requires either processing logs in FIFO order out of the dirty log (to
+roll up the collections to compact), or somehow compacting the data.  The following techniques give
+sufficient generality to handle every case we need:
 
-To compromise, we want the following:
-- Tracking dirty per collection _written_, not _existing_.
-- Partitioning over dirty collections.
-- A way to support more collections in the future should costs be too great.
+- Forcibly compact things that are at the head of the log.
+- Re-write dirty log entries at the end of the log so they can be collected from the beginning.
 
-```text
-wal3/dirty/<prefix bits>/<suffix bits>
-```
+To facilitate this, we will store the reinsertion count and initial insertion time as well as chroma
+collection ID when inserting the dirty log entry.  This allows us to roll up the dirty log by simply
+reading the first N records, picking those that are too old, picking those that are too big, and
+then reinsert any that are not selected by this algorithm.
 
-The number of prefix bits must be configurable, but shouldn't change much in practice.  Should it
-change, it is a straw-man away from being recreated.
+Thus each wal3 log service will independently manage its own dirty log.  This allows us to scale
```

Comment:
Maybe worth talking about what happens to dirty logs as we scale up /down services